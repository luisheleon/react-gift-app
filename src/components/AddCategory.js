import React, { useState } from "react";
import PropTypes from "prop-types";
import { Button, TextField } from "@mui/material";

const AddCategory = ({ setCategories }) => {
  const [errorText, setErrorText] = useState(false);
  const [value, setValue] = useState("");

  const handleAddCategorie = (e) => {
    e.preventDefault();
    if (value.length) {
      setCategories((catego) => {
        if (!catego.includes(value)) {
          return [value, ...catego];
        } else {
          setErrorText(true);
          console.log("Elemento ya fue agregado");
          return catego;
        }
      });
      setValue("");
    }
  };
  return (
    <React.Fragment>
      <h2 id="subtitle">Add category</h2>
      <form onSubmit={handleAddCategorie}>
        <div className="newCategoryForm">
          <TextField
            required
            error={errorText}
            id="outlined-required"
            label="Add Category"
            value={value}
            onChange={(e) => {
              e.target.value.length && setErrorText(false);
              setValue(e.target.value);
            }}
            size="small"
          />
          {errorText && <label className="labelError">El elemento ya fue agregado</label>}
        </div>

        <Button id="btnAddCategory" type="submit" variant="outlined" size="medium" disabled={!value.length}>
          Add
        </Button>
      </form>
    </React.Fragment>
  );
};

AddCategory.propType = {
  handleNewCategory: PropTypes.func.isRequired,
};

export default AddCategory;
