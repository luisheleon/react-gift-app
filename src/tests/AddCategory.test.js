import AddCategory from "../components/AddCategory";
import { shallow, mount } from "enzyme";
import TextField from "@mui/material/TextField";

describe("Test <AddCategory />", () => {
  const handleChange = jest.fn();
  let wrapper;
  const eventSubmit = {
    preventDefault() {},
  };
  const event = {
    target: {
      value: "hola",
    },
  };
  beforeEach(() => {
    jest.clearAllMocks();
    wrapper = shallow(<AddCategory setCategories={handleChange} />);
  });
  test("should return snapshop", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should update component when input is changed", () => {
    wrapper.find(TextField).at(0).simulate("change", event);
    expect(wrapper.find("p").text().trim()).toBe(event.target.value);
  });

  test("should prevent submit if the input is empty", () => {
    const eventSubmit = {
      preventDefault() {},
    };
    wrapper.find("form").simulate("submit", eventSubmit);
    expect(handleChange).not.toHaveBeenCalled();
  });

  test("should submit if the input is not empty", () => {
    wrapper.find(TextField).at(0).simulate("change", event);
    wrapper.find("form").simulate("submit", eventSubmit);
    expect(handleChange).toHaveBeenCalled();
  });

  test("should return input categories to empty", () => {
    wrapper.find(TextField).simulate("change", event);
    expect(wrapper.find(TextField).props().value).toBe(event.target.value);
    wrapper.find("form").simulate("submit", eventSubmit);
    const eventEmpty = { ...event };
    eventEmpty.target.value = "";
    wrapper.find(TextField).simulate("change", eventEmpty);
    expect(handleChange).toHaveBeenCalled();
    expect(wrapper.find(TextField).props().value).toBe("");
  });
});
