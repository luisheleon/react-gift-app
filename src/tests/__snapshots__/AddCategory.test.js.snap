// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <AddCategory /> should return snapshop 1`] = `
<Fragment>
  <h2
    id="subtitle"
  >
    Add category
  </h2>
  <form
    onSubmit={[Function]}
  >
    <div
      className="newCategoryForm"
    >
      <ForwardRef(TextField)
        className="inputCategory"
        error={false}
        id="outlined-required"
        label="Add Category"
        onChange={[Function]}
        required={true}
        size="small"
        value=""
      />
    </div>
    <p />
    <ForwardRef(Button)
      disabled={true}
      id="btnAddCategory"
      size="medium"
      type="submit"
      variant="outlined"
    >
      Add
    </ForwardRef(Button)>
  </form>
</Fragment>
`;
